File: piggybank/pig.pas
program pig;
var
   T : longint ; //casos a evaluar
   counter : longint ; //contador de ciclo
   counter1 : longint ; //contador de ciclo interno
   E : longint ; //peso vacio
   F : longint ; //peso lleno
   N : longint ; //catidad de monedas diferentes
   P : array of longint;
   W : array of longint;
   R : array of real; //relacion valor peso
   min : longint ;//posicion de moneda a usar
   comp : real ;
   listo : boolean ; //ya se termino
   diferencia : longint ;//peso de monedas del cerdo
   existsol : boolean ; //existe o no solucion
   suma : longint ;
   minmoney : longint ;
   
   
begin
   listo:=false;
   suma:=0;
   counter;=0;
   readln(T);
   repeat begin
      readln(E,F);//peso vacio-lleno
	  diferencia:=(F - E);
      if not ((1 <= E) and (E <= F) and (F <= 10000)) THEN begin
         halt;
      end;
	  readln(N);//cantidad de monedas
      if not ((1 <= N) and (N <= 500)) THEN begin
         halt;
      end;
	  setlength(P,N);//valor moneda
	  setlength(W,N);//peso moneda
	  counter1:=0;
	  repeat begin
	     readln(P[counter1],W[counter1]);
		 R[counter1]:=(P[counter1] / W[counter1])
	     counter1:=(counter1 + 1);
	  end until (counter1 = N);
	  counter1:=0;
	  existsol:=false;
	  repeat begin
	     //chequeando mas pequeno
		  if (counter1 = 0) then begin
			 comp:=R[0];
			 min:=0;
			 repeat begin
			    if (R[(counter1 + 1)] < comp) then begin
			 	  comp:=R[(counter1 + 1)];
			      min:=(counter1 + 1);
			 	end;
			    counter1:=(counter1 + 1);
		     end until (counter1 = (N - 1));
		  end else begin
		     R[min]:=(-1);
			 counter1:=0;
			 repeat begin
			     if (counter1 < N) then begin
				     if (R[counter1] <> -1) then begin
					     comp:=R[counter1];
						 min:=counter1;
						 listo:=true;
					 end;
				 end;
				 if (counter1 = N) THEN begin
				     existsol:=false;
			     end;
				 counter1:=(counter1 + 1);
			 end until ((counter1 > N) or (listo = true));
			 listo:=false;
			 counter1:=1;
			 if (existsol <> 0) then begin
			     counter1:=0;
			     repeat begin
			         if ((R[counter1] < comp) and (R[counter1] <> -1)) then begin
			        	  comp:=R[(counter1 + 1)];
			              min:=(counter1 + 1);
			         end;
			         counter1:=(counter1 + 1);
		          end until (counter1 = N);
				  counter1:=1;
			 end;
		 end;
		 //evaluando cantidad de monedas
		 if (existsol = false) then begin
		     writeln('This is impossible.');
			 listo:=true;
		 end else begin
		     suma:=(suma + ((
			 minmoney
	     
	  end until (listo = true);
	  suma:=0;
	  listo:=false;
	  existsol:=true;
	  counter:=(counter + 1);
   end until (counter = T);
   
   
   
   

File: piggybank/pig.cpp
#include <iostream>
using namespace std;
long T, E, F, N, diferencia;
int* a = NULL;
int* b = NULL;
int* c = NULL;

int valmin (long X, long Y, long z, long v, long w, long l)
   {
      //valmin(X) = min(B[counter1] + valmin (X - C[counter1]))
      if (X == 0) {
         return (0);
      } else {
         if (a[X] == -1) {
      for ( z=0 ; z!=N ; z++){
         if (((X - (c[z]))) >= 0) {
             v = valmin((X - (c[z])),(Y + 1),0,0,0,-1);
             if (v >= 0) {
                 w = ((b[z]) + v);
                 if (l == (-1)) {
                      l = w;
                 } else {
                    if (w < l) {
                     l = w;
                    };
                 };
             };
         };
      };
      if (l == -1) {
         l = -2;
      };
      a[X] = l;
      return (l);
         } else {
             return (a[X]);
         };
      };
   };

int main ()
{
   cin >> T ;
      for (int K = 0 ; K!=T ; K++) {
      cin >> E >> F;//peso vacio-lleno
	  diferencia=(F - E);
	  cin >> N;//cantidad de monedas
      b = new int[N];
      c = new int[N]; 
      a = new int[(diferencia + 1)];
      for (int I = 0 ; I<=diferencia ; I++) a[I]=-1;
      for (int I = 0 ; I!=N ; I++) {cin >> b[I] >> c[I] ;}; //valor - peso
      if (diferencia != 0) {
         if (valmin(diferencia,0,0,0,0,-1) == -2) {
             cout << "This is impossible." << endl;
         } else {
             cout << "The minimum amount of money in the piggy-bank is " << a[diferencia] << "." << endl ;
         }
      } else {
         cout << "The minimum amount of money in the piggy-bank is o." << endl;
      };
     delete [] a; 
     a = NULL; 
     delete [] b; 
     b = NULL; 
     delete [] c; 
     c = NULL; 
   };
}